File: tests/09-test_update_session.ts
import * as anchor from "@anchor-lang/core";
import { Program } from "@anchor-lang/core";
import {
  ComputeBudgetProgram,
  Keypair,
  PublicKey,
  TransactionInstruction,
} from "@solana/web3.js";
import assert from "assert";

import { R3SessionKeys } from "../target/types/r3_session_keys";
import { MockProgram, mockProgramIdl } from "./fixtures/mock-program";
import * as mock from "./fixtures/mock-program";
import {
  DUMMY_ANCHOR_DISCRIMINATOR,
  TARGET_PROGRAM_PLACEHOLDER,
} from "./utils/constants";
import {
  ensureProgramInitialized,
  executeWithSession,
  loadProgramFromFile,
  sendExpectError,
  setAccountData,
  timestampFromFuture,
} from "./utils/helpers";
import * as sessionKeys from "./utils/pda";

describe("09 - Update session", () => {
  const provider = anchor.AnchorProvider.env();
  anchor.setProvider(provider);
  const baseWallet = provider.wallet as anchor.Wallet;
  const connection = provider.connection;
  const program = anchor.workspace.r3SessionKeys as Program<R3SessionKeys>;

  let mockProgram: Program<MockProgram>;
  let mockCounterPda: PublicKey;

  before(async () => {
    await ensureProgramInitialized(program, baseWallet);
    await loadProgramFromFile(
      connection,
      mock.MOCK_PROGRAM_ID,
      mock.MOCK_PROGRAM_SO_PATH
    );
    mockProgram = new Program(
      mockProgramIdl as MockProgram,
      provider
    ) as Program<MockProgram>;
    mockCounterPda = mock.deriveCounterPda(mockProgram.programId);
  });

  async function initializeMockCounter(authority: PublicKey) {
    const counterAccount = await connection.getAccountInfo(mockCounterPda);
    if (!counterAccount) {
      await mockProgram.methods
        .initialize()
        .accounts({ payer: baseWallet.publicKey })
        .signers([baseWallet.payer])
        .rpc();
    }

    await setAccountData(
      connection,
      mockCounterPda,
      mock.encodeCounterAccount(0, authority)
    );
  }

  async function approveSession(
    smartWalletOwner: Keypair,
    sessionKey: PublicKey,
    makeTransactionUnique = false
  ) {
    const builder = program.methods
      .approveSession(sessionKey)
      .accounts({ smartWalletOwner: smartWalletOwner.publicKey })
      .signers([baseWallet.payer, smartWalletOwner]);

    if (makeTransactionUnique) {
      builder.preInstructions([
        ComputeBudgetProgram.setComputeUnitLimit({ units: 200_001 }),
      ]);
    }

    await builder.rpc();
  }

  function incrementInstruction(
    authority: PublicKey
  ): Promise<TransactionInstruction> {
    return mockProgram.methods
      .increment()
      .accounts({ authority })
      .instruction();
  }

  it("resets approval until the owner approves the changed grant", async () => {
    const smartWalletOwner = Keypair.generate();
    const sessionKey = Keypair.generate();
    const userSmartWallet = sessionKeys.deriveUserSmartWalletPda(
      program.programId,
      smartWalletOwner.publicKey
    );
    const session = sessionKeys.deriveSessionPda(
      program.programId,
      userSmartWallet,
      sessionKey.publicKey
    );

    await program.methods
      .createSmartWallet(smartWalletOwner.publicKey)
      .accounts({ admin: baseWallet.publicKey })
      .signers([baseWallet.payer])
      .rpc();
    await initializeMockCounter(userSmartWallet);

    const expiresAtT1 = await timestampFromFuture(connection);
    const expiresAtT2 = expiresAtT1.add(new anchor.BN(1_000));
    await program.methods
      .createSession(
        sessionKey.publicKey,
        TARGET_PROGRAM_PLACEHOLDER,
        expiresAtT1,
        DUMMY_ANCHOR_DISCRIMINATOR,
        DUMMY_ANCHOR_DISCRIMINATOR.length
      )
      .accounts({
        sessionExecutor: baseWallet.publicKey,
        userSmartWallet,
      })
      .signers([baseWallet.payer])
      .rpc();
    await approveSession(smartWalletOwner, sessionKey.publicKey);

    await program.methods
      .updateSession(
        sessionKey.publicKey,
        mockProgram.programId,
        expiresAtT2,
        mock.INCREMENT_DISCRIMINATOR,
        mock.INCREMENT_DISCRIMINATOR.length
      )
      .accounts({
        sessionExecutor: baseWallet.publicKey,
        userSmartWallet,
      })
      .signers([baseWallet.payer])
      .rpc();

    const updated = await program.account.session.fetch(session);
    assert.equal(
      updated.sessionExecutor.toBase58(),
      baseWallet.publicKey.toBase58()
    );
    assert.equal(
      updated.sessionKey.toBase58(),
      sessionKey.publicKey.toBase58()
    );
    assert.equal(
      updated.targetProgram.toBase58(),
      mockProgram.programId.toBase58()
    );
    assert.ok(updated.expiresAt.eq(expiresAtT2));
    assert.deepEqual(
      Buffer.from(updated.allowedInstructionsDiscriminators),
      mock.INCREMENT_DISCRIMINATOR
    );
    assert.equal(
      updated.discriminatorSize,
      mock.INCREMENT_DISCRIMINATOR.length
    );
    assert.ok("waitingForApproval" in updated.status);

    const incrementIx = await incrementInstruction(userSmartWallet);
    const error = await sendExpectError(
      executeWithSession(program, {
        sessionExecutor: baseWallet.publicKey,
        sessionKey: sessionKey.publicKey,
        userSmartWallet,
        targetInstruction: incrementIx,
      })
        .signers([baseWallet.payer, sessionKey])
        .rpc()
    );
    assert.ok(error.includes("SessionNotApproved"), error);

    await approveSession(smartWalletOwner, sessionKey.publicKey, true);
    await executeWithSession(program, {
      sessionExecutor: baseWallet.publicKey,
      sessionKey: sessionKey.publicKey,
      userSmartWallet,
      targetInstruction: incrementIx,
    })
      .signers([baseWallet.payer, sessionKey])
      .rpc();

    const counter = await mockProgram.account.counter.fetch(mockCounterPda);
    assert.equal(counter.count.toNumber(), 1);
    assert.equal(counter.authority.toBase58(), userSmartWallet.toBase58());
  });
});

File: programs/r3-session-keys/tests/09-test_update_session.rs
mod common;

use {
    anchor_lang::AccountSerialize,
    common::{
        client, load, mock_client, send_tx_expect_error, send_tx_expect_ok, timestamp_from_future,
        Env, DUMMY_ANCHOR_DISCRIMINATOR, TARGET_PROGRAM_PLACEHOLDER,
    },
    mock_client::mock_program::accounts::Counter,
    r3_session_keys::state::session::{Session, SessionStatus},
    solana_keypair::Keypair,
    solana_signer::Signer,
};

fn load_mock_program(env: &mut Env) {
    let bytes = include_bytes!(concat!(
        env!("CARGO_MANIFEST_DIR"),
        "/../../tests/fixtures/mock-program/mock_program.so"
    ));
    env.svm
        .add_program(mock_client::mock_program::ID, bytes)
        .unwrap();
}

fn initialize_mock_counter(
    env: &mut Env,
    mock_client: &mock_client::MockClient,
    authority: anchor_lang::prelude::Pubkey,
) {
    let counter = mock_client::counter_pda();
    let initialize_ix = mock_client.initialize(env.admin.pubkey());
    send_tx_expect_ok(&mut env.svm, initialize_ix, &[&env.admin]);

    let mut counter_account = env.svm.get_account(&counter).unwrap();
    let counter_state = Counter {
        count: 0,
        authority,
    };
    let mut account_data = counter_account.data.as_mut_slice();
    counter_state.try_serialize(&mut account_data).unwrap();
    env.svm.set_account(counter, counter_account).unwrap();
}

#[test]
fn test_update_session_resets_approval() {
    let client = client::R3SessionKeysClient::new();
    let mut env = Env::new();
    let initialize_ix = client.initialize(env.admin.pubkey());
    send_tx_expect_ok(&mut env.svm, initialize_ix, &[&env.admin]);

    load_mock_program(&mut env);
    let mock_client = mock_client::MockClient::new();
    let smart_wallet_owner = Keypair::new();
    let session_key = Keypair::new();
    let (create_wallet, user_smart_wallet, _) =
        client.create_smart_wallet(env.admin.pubkey(), smart_wallet_owner.pubkey());
    send_tx_expect_ok(&mut env.svm, create_wallet, &[&env.admin]);
    initialize_mock_counter(&mut env, &mock_client, user_smart_wallet);

    let expires_at_t1 = timestamp_from_future();
    let expires_at_t2 = expires_at_t1 + 1000;
    let discriminator_set_a = DUMMY_ANCHOR_DISCRIMINATOR.to_vec();
    let discriminator_set_b = mock_client::increment_discriminator().to_vec();
    let (create_session, session, _) = client.create_session(
        env.admin.pubkey(),
        user_smart_wallet,
        session_key.pubkey(),
        TARGET_PROGRAM_PLACEHOLDER,
        expires_at_t1,
        discriminator_set_a,
        8,
    );
    send_tx_expect_ok(&mut env.svm, create_session, &[&env.admin]);

    let approve_session = client.approve_session(
        smart_wallet_owner.pubkey(),
        user_smart_wallet,
        session_key.pubkey(),
    );
    send_tx_expect_ok(
        &mut env.svm,
        approve_session,
        &[&env.admin, &smart_wallet_owner],
    );

    let update_session = client.update_session(
        env.admin.pubkey(),
        user_smart_wallet,
        session_key.pubkey(),
        mock_client::mock_program::ID,
        expires_at_t2,
        discriminator_set_b.clone(),
        u8::try_from(discriminator_set_b.len()).unwrap(),
    );
    send_tx_expect_ok(&mut env.svm, update_session, &[&env.admin]);

    let updated: Session = load(&env, &session);
    assert_eq!(updated.session_executor, env.admin.pubkey());
    assert_eq!(updated.session_key, session_key.pubkey());
    assert_eq!(updated.target_program, mock_client::mock_program::ID);
    assert_eq!(updated.expires_at, expires_at_t2);
    assert_eq!(
        updated.allowed_instructions_discriminators,
        discriminator_set_b
    );
    assert_eq!(updated.discriminator_size, 8);
    assert_eq!(updated.status, SessionStatus::WaitingForApproval);

    let increment_ix = mock_client.increment(user_smart_wallet);
    let execute = client.execute_with_session(
        env.admin.pubkey(),
        session_key.pubkey(),
        user_smart_wallet,
        increment_ix.clone(),
    );
    let error = send_tx_expect_error(&mut env.svm, execute, &[&env.admin, &session_key]);
    assert!(error.contains("Error Code: SessionNotApproved"), "{error}");

    let approve_session = client.approve_session(
        smart_wallet_owner.pubkey(),
        user_smart_wallet,
        session_key.pubkey(),
    );
    send_tx_expect_ok(
        &mut env.svm,
        approve_session,
        &[&env.admin, &smart_wallet_owner],
    );

    let execute = client.execute_with_session(
        env.admin.pubkey(),
        session_key.pubkey(),
        user_smart_wallet,
        increment_ix,
    );
    send_tx_expect_ok(&mut env.svm, execute, &[&env.admin, &session_key]);

    let counter: Counter = load(&env, &mock_client::counter_pda());
    assert_eq!(counter.count, 1);
    assert_eq!(counter.authority, user_smart_wallet);
}

#[test]
fn test_update_rejects_non_executor() {
    let client = client::R3SessionKeysClient::new();
    let mut env = Env::new();
    let initialize_ix = client.initialize(env.admin.pubkey());
    send_tx_expect_ok(&mut env.svm, initialize_ix, &[&env.admin]);

    let smart_wallet_owner = Keypair::new();
    let session_key = Keypair::new();
    let wrong_executor = Keypair::new();
    let (create_wallet, user_smart_wallet, _) =
        client.create_smart_wallet(env.admin.pubkey(), smart_wallet_owner.pubkey());
    send_tx_expect_ok(&mut env.svm, create_wallet, &[&env.admin]);
    let (create_session, _, _) = client.create_session(
        env.admin.pubkey(),
        user_smart_wallet,
        session_key.pubkey(),
        TARGET_PROGRAM_PLACEHOLDER,
        timestamp_from_future(),
        DUMMY_ANCHOR_DISCRIMINATOR.to_vec(),
        8,
    );
    send_tx_expect_ok(&mut env.svm, create_session, &[&env.admin]);

    let update_session = client.update_session(
        wrong_executor.pubkey(),
        user_smart_wallet,
        session_key.pubkey(),
        TARGET_PROGRAM_PLACEHOLDER,
        timestamp_from_future(),
        DUMMY_ANCHOR_DISCRIMINATOR.to_vec(),
        8,
    );
    let error = send_tx_expect_error(&mut env.svm, update_session, &[&env.admin, &wrong_executor]);
    assert!(
        error.contains("Error Code: UnauthorizedSessionExecutor"),
        "{error}"
    );
}

#[test]
fn test_update_rejects_revoked_session() {
    let client = client::R3SessionKeysClient::new();
    let mut env = Env::new();
    let initialize_ix = client.initialize(env.admin.pubkey());
    send_tx_expect_ok(&mut env.svm, initialize_ix, &[&env.admin]);

    let smart_wallet_owner = Keypair::new();
    let session_key = Keypair::new();
    let (create_wallet, user_smart_wallet, _) =
        client.create_smart_wallet(env.admin.pubkey(), smart_wallet_owner.pubkey());
    send_tx_expect_ok(&mut env.svm, create_wallet, &[&env.admin]);
    let (create_session, _, _) = client.create_session(
        env.admin.pubkey(),
        user_smart_wallet,
        session_key.pubkey(),
        TARGET_PROGRAM_PLACEHOLDER,
        timestamp_from_future(),
        DUMMY_ANCHOR_DISCRIMINATOR.to_vec(),
        8,
    );
    send_tx_expect_ok(&mut env.svm, create_session, &[&env.admin]);

    let revoke_session = client.revoke_session(
        smart_wallet_owner.pubkey(),
        user_smart_wallet,
        session_key.pubkey(),
    );
    send_tx_expect_ok(
        &mut env.svm,
        revoke_session,
        &[&env.admin, &smart_wallet_owner],
    );

    let update_session = client.update_session(
        env.admin.pubkey(),
        user_smart_wallet,
        session_key.pubkey(),
        TARGET_PROGRAM_PLACEHOLDER,
        timestamp_from_future(),
        DUMMY_ANCHOR_DISCRIMINATOR.to_vec(),
        8,
    );
    let error = send_tx_expect_error(&mut env.svm, update_session, &[&env.admin]);
    assert!(
        error.contains("Error Code: InvalidSessionStatus"),
        "{error}"
    );
}
